### PR TITLE
Search indexing (Part 1) - Docs restructure

### PR DESCRIPTION
**Pull request by fonty1**
**Part 1 - Restructure**
It's long but it's almost entirely text replacements. A quick sanity check is fine.

A Docs exit (Dexit) of the docs site map object (formerly `navLinks`) now called `docsMap` (`$docs/docsMap.js`). `docsMap` is what the indexer, docs navigation and page turner components digest to do their job. I also extracted the raw docs links from the routes definition file and put them in `docsLinks` (`$shared/../links`). These were necessary steps due to the needs of the indexing script.

The structure of the `docsMap` includes some 'convenience properties' for the indexer to chew on which are important for generating the all the information required for each search result (https://share.goabstract.com/53140352-e4a1-4755-a811-84d464cc6d50?mode=design&sha=cf9b31f3cb25734146d94b7f47ce8f0807427d82). 

I also added:
- API explorer to be a top level section
- SDK split into 3 separate pages (as per design)
- DU docs flag from `DATA_UNIONS` to `DATA_UNIONS_DOCS`

**Part 2 - Indexer**
- The actual node indexer script.
- Handle docs page link/duplication

**Part 3 - CI & loose ends**
- Integrating the node script into CI
- Tests for robustness
- Add guidelines/docs for how to add to the docs and keep the indexer happy

### CODE REVIEW

**Comment by juhah** [comment]
There are some case-sensitive paths:
Screen Shot 2020-04-30 at 10 30 04
Screen Shot 2020-04-30 at 10 24 36

This works (`Sdks` -> `SDKs`):
Screen Shot 2020-04-30 at 10 30 20

File: app/src/app/Docs.jsx [hunk 1]
```diff
@@ -3,210 +3,203 @@
...
+    <Redirect exact from={links.api} to={links.apiOverview} key="ApiOverview" />,
+    <Route exact path={links.authentication} component={Authentication} key="Authentication" />,
+    <Route exact path={links.apiStreamsViaApi} component={UsingStreamsViaApi} key="usingStreamsViaApi" />,
+    <Route exact path={links.apiExplorer} component={ApiExplorer} key="apiExplorer" />,
     // Technical Notes routes
-    <Route exact path={docs.technicalNotes.root} component={TechnicalNotesDocsPage} key="technicalNotes" />,
+    <Route exact path={links.technicalNotes} component={TechnicalNotes} key="technicalNotes" />,
     // Docs Root
-    <Redirect exact from={docs.main} to={docs.introduction} key="DocsRoot" />,
+    <Redirect exact from="/docs" to={links.introduction} key="DocsRoot" />,
```
Comment (juhah) [inline app/src/app/Docs.jsx:202]: why hardcoded url here?

**Comment by fonty1** [comment]
> There are some case-sensitive paths:


Ah that's why its working locally and failing here! Will address. Thanks

File: app/test/unit/app/Docs.test.jsx [hunk 2]
```diff
@@ -216,10 +216,37 @@ jest.mock('$docs/components/DocsPages/Marketplace/IntroToMarketplace', () => ({
         </div>
     ),
 }))
-jest.mock('$docs/components/DocsPages/SDKs', () => ({
+jest.mock('$docs/components/DocsPages/Sdks/SdkOverview', () => ({
```
Comment (juhah) [inline app/test/unit/app/Docs.test.jsx:219]: Casing here as well:
Screen Shot 2020-04-30 at 10 45 19

File: app/src/app/Docs.jsx (hunk 1 above)
Comment (fonty1) [reply app/src/app/Docs.jsx]: Must have been a copy/paste error, fixed now.

**Comment by fonty1** [comment]
@juhah please take another look. Hopefully Git saw all those case changes. https://github.com/streamr-dev/streamr-platform/pull/983/commits/d11bb645db3f3d5a9fba83c90ed3ffe20e92c298

I did a general consistency check to avoid these sorts of bugs in the future - 
SDK -> Sdk
sdks -> sdk
API -> Api

**Comment by fonty1** [comment]
I'm having issues switching between branches, so it seems git can't handle pure casing changes that well. I will fix that now.

**Comment by juhah** [comment]
@fonty1 Was this the error that you were referring to:

Screen Shot 2020-05-04 at 11 45 34

Screen Shot 2020-05-04 at 11 45 38

The folder names are `DataUnionsSDK` and `DataUnionsCore` (missing `In`).

**Comment by fonty1** [comment]
> @fonty1 Was this the error that you were referring to:
> The folder names are `DataUnionsSDK` and `DataUnionsCore` (missing `In`).

Sort of :) At least for me, filename casing changes (SDK -> sdk) do not get recognised on their own, so I needed to change their names to get the casing update to register otherwise we run into problems.. then I missed some references inbetween. 

I changed `DocsPages` to `Pages` and removed the `in` from content paths in a roundabout way to get past this.

**Comment by fonty1** [comment]
@juhah all clear now

File: app/src/docs/content/gettingStarted/getBuilding.mdx [hunk 3]
```diff
@@ -1,24 +1,24 @@
...
+import { docsLinks } from '$shared/../links'
 
 ## Get building
 Here are some ideas to help you start building on the Streamr Platform.
 
 ### Tutorials
-- <Link to={links.docs.tutorials.buildingPubSub}>Build a simple pub/sub system</Link>
-- <Link to={links.docs.tutorials.buildingCustomModule}>Build a custom canvas module</Link>
+- <Link to={docsLinks.buildingPubSub}>Build a simple pub/sub system</Link>
+- <Link to={docsLinks.buildingCustomModule}>Build a custom canvas module</Link>
```
Comment (mondoreale) [inline app/src/docs/content/gettingStarted/getBuilding.mdx:9]: Updated required:

`docsLinks.buildingPubSub` → `docsLinks.buildingPubSubTutorial`
`buildingCustomModule ` → `docsLinks.customModuleTutorial`

**Comment by mondoreale** [comment]
I also hear @juhah. I'd love to see `routes` in current or improved version being used across the app. I find it quite handy. [`PLATFORM-1360`](https://streamr.atlassian.net/browse/PLATFORM-1360) is all about the migration.

**Comment by fonty1** [comment]
Thanks @juhah @mondoreale those bugs are fixed. 

I tried out the indexer again and realised I had to separate the `docsLinks` even further, now into its own file in https://github.com/streamr-dev/streamr-platform/pull/983/commits/93aa9211924f07d51bdc72428050f8cf69d072b8. Maybe that makes ticket [PLATFORM-1360](https://streamr.atlassian.net/browse/PLATFORM-1360) easier I hope!

I'll merge this now, ping me if anything doesn't fit right.

**Pull request merged**